File: frameworks/java/spring-boot/UnitTests/src/test/java/unittests/HeadersTests.java
package unittests;

import java.net.http.HttpRequest;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import tools.jackson.databind.node.ObjectNode;

class HeadersTests extends SpringApp {

    // rb:test headers.few,headers.many
    @ParameterizedTest
    @Tag("headers.few")
    @Tag("headers.many")
    @ValueSource(ints = {0, 25})
    void headersNothingReadsLeaveTheAnswerAlone(int unread) throws Exception {
        Answer.is(Expected.json("items.small.json"), send(bound(request("/headers"), unread)));
    }

    // rb:test headers.bind_few,headers.bind_many
    @ParameterizedTest
    @Tag("headers.bind_few")
    @Tag("headers.bind_many")
    @ValueSource(ints = {0, 25})
    void threeHeadersAreBoundAndEchoedOneAsAnInteger(int unread) throws Exception {
        ObjectNode echo = Expected.JSON.createObjectNode().put("tenant", "qwertyuiopas").put("requestId", "0123456789abcdef").put("account", 482913);

        Answer.is(Expected.withEcho("items.small.json", echo), send(bound(request("/headers/bind"), unread)));
    }

    /**
     * The three headers the binding rows bind, and as many more as asked that nothing reads. The
     * many rows send twenty-five of those.
     */
    private static HttpRequest.Builder bound(HttpRequest.Builder request, int unread) {
        request.header("x-rb-tenant", "qwertyuiopas").header("x-rb-request-id", "0123456789abcdef").header("x-rb-account", "482913");
        for (int i = 0; i < unread; i++) {
            request.header("x-rb-unread-" + i, "unread");
        }
        return request.GET();
    }
}

File: frameworks/dotnet/minimal-apis/UnitTests/HeadersTests.cs
namespace UnitTests;

public sealed class HeadersTests(MinimalApisApp app) : IClassFixture<MinimalApisApp>
{
    // rb:test headers.few,headers.many
    [Theory]
    [Trait("corpus", "headers.few")]
    [Trait("corpus", "headers.many")]
    [InlineData(0)]
    [InlineData(25)]
    public async Task Headers_nothing_reads_leave_the_answer_alone(int unread)
    {
        using HttpRequestMessage request = Bound(new(HttpMethod.Get, "/headers"), unread);

        using HttpResponseMessage response = await app.CreateClient().SendAsync(request);

        await Answer.Is(Expected.Json("items.small.json"), response);
    }

    // rb:test headers.bind_few,headers.bind_many
    [Theory]
    [Trait("corpus", "headers.bind_few")]
    [Trait("corpus", "headers.bind_many")]
    [InlineData(0)]
    [InlineData(25)]
    public async Task Three_headers_are_bound_and_echoed_one_as_an_integer(int unread)
    {
        using HttpRequestMessage request = Bound(new(HttpMethod.Get, "/headers/bind"), unread);

        using HttpResponseMessage response = await app.CreateClient().SendAsync(request);

        JsonObject echo = new() { ["tenant"] = "qwertyuiopas", ["requestId"] = "0123456789abcdef", ["account"] = 482913 };
        await Answer.Is(Expected.WithEcho("items.small.json", echo), response);
    }

    /// <summary>
    /// The three headers the binding rows bind, and as many more as asked that nothing reads.
    /// The many rows send twenty-five of those.
    /// </summary>
    private static HttpRequestMessage Bound(HttpRequestMessage request, int unread)
    {
        request.Headers.Add("x-rb-tenant", "qwertyuiopas");
        request.Headers.Add("x-rb-request-id", "0123456789abcdef");
        request.Headers.Add("x-rb-account", "482913");
        for (int i = 0; i < unread; i++)
        {
            request.Headers.Add($"x-rb-unread-{i}", "unread");
        }
        return request;
    }
}
